src/dtype.rs

```diff
@@ -57,6 +58,22 @@ pub fn dtype<T: Element>(py: Python) -> &PyArrayDescr {
 }
 
 impl PyArrayDescr {
+    /// Creates a new dtype object from an arbitrary object.
+    ///
+    /// Equivalent to invoking the constructor of `np.dtype`.
+    pub fn new<'py>(py: Python<'py>, obj: &(impl ToPyObject + ?Sized)) -> PyResult<&'py Self> {
```

Comment:
I am not a fan of the overly magical type signature here. If this takes `PyObject`, I'd suggest using this for the type signature. This would also avoid making the method generic.

But even if we do keep it generic, I think that it should not use impl trait in argument position but a named type parameter and a bound as we prevent specifying the type via turbofish notation otherwise. But as written above, I don't think making this generic carries its weight in the first place.

Reply:
The generic signature is to allow e.g. `::new(py, "float64")` and the like without extra fuss - since the typical valid arguments for `np.dtype()` looks quite simple - strings, tuples, lists of strings and tuples etc which can be represented with Rust builtins. This falls inline with many signatures in pyo3 itself, e.g. to just pick a a random one - [`PySet::new`](https://docs.rs/pyo3/latest/pyo3/types/struct.PySet.html#method.new) which looks almost exactly the same, `impl` part and the slice part aside.

The `impl` can be removed, in which case it'll be like
```rust
pub fn new<'py, T: ToPyObject>(py: Python<'py>, obj: &T) -> PyResult<&'py Self> {}
```

Reply:
In this case, please split this into a
```rust
#[inline]
pub fn new<'py, T: ToPyObject>(py: Python<'py>, obj: &T) -> PyResult<&'py Self> {
  Self::new_impl(py, obj.to_object(py))
}
```
and a private
```rust
fn new_impl<'py,>(_py: Python<'py>, obj: PyObject) -> PyResult<&'py Self> { ... }
```
to keep the generic code as limited as possible. (I am not invested in the names, just the first thing that came to mind.)

Reply:
(Or a nested function if you prefer.)

Reply:
I've force-pushed the update splitting it into two functions.

(To be honest, not sure I entirely follow, since the amount of generic code is the same, public methods are the same, and in the original function all code was non-generic except the first line that shadowed a generic input argument with non-generic one. But maybe I'm misunderstanding something)

Reply:
The `#[inline]` should be on the public method, i.e. the `.to_object(py)` must be generic and should happen directly at the call site. The actual method does not need to be and may benefit from increased code density if generated out-of-line.

The overall amount of generic is reduced as the all invocations of the public method share a single instantiation of the private method as it is not generic any more. The compiler might decide to inline this, but at least we are giving it a chance to not do that and we are avoiding instantiating that inner code repeatedly for each type argument even though it does not actually depend on it (which is what is happening without the split).

Reply:
https://doc.rust-lang.org/stable/src/alloc/raw_vec.rs.html#365 is probably a good example where the technique is explained in more detail.

Reply:
Oh, it's all from the codegen perspective - as I said, I was misunderstanding the point, that makes sense indeed 👍

(moved the `#[inline]` to the outer method)